Support for new velux api, added cover.velux

Julius2342 opened the pull request:
## Description:

Upgraded pyvlx to version 0.2.*. Users will need to upgrade to a recent firmware version. Old and new firmwareversions are not compatible (both use a completely different API).

Added support for controlling Velux windows one by one.

Configuration should stay the same. Component will fetch the available devices/nodes from API.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/7867


*EDIT:* **Within the changelog of the release we should point to the page where the new Firmware may be downloaded: https://www.velux.com/api/klf200**

Julius2342 committed:
Support for new velux api, added cover.velux

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into new-velux-api

Julius2342 committed:
More steps on new velux covers.

Julius2342 committed:
Merge branch 'dev' into new-velux-api

WoodsterDK commented:
Hi I noticed one issue:

    async def async_open_cover(self, **kwargs):
        """Open the cover."""
        await self.node.close()

should be     async def async_open_cover(self, **kwargs):
        """Open the cover."""
        await self.node.open()

And 

    def current_cover_position(self):
        """Return the current position of the cover."""
        return 0

can be changed to support position query:

    def current_cover_position(self):
        """Return the current position of the cover."""
        return self.node.current_position

Julius2342 commented:
@WoodsterDK ty for you feedback! i already spotted the `s/close/open/` - thing, but did not yet commit. I'm currently working on retrieving the current position from various feedback/status messages, but this still takes some iterations. I will post an update here if i have a working prototype :)

WoodsterDK commented:
@Julius2342 no problem - I'm up for testing it, when you have something ready.
Let me know if you need some help.

Julius2342 committed:
correct position handling of velux windows

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into new-velux-api

Julius2342 committed:
Following suggestion from hound.

Julius2342 commented:
@WoodsterDK ^^

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into new-velux-api

WoodsterDK commented:
Hi,
The covers are not being inserted as entities.
- I think there might be an issue if no scenes are present
- I have seen an "out of index" 
![image](https://user-images.githubusercontent.com/12181717/50217801-19849400-038a-11e9-9273-918c1222f765.png)

Julius2342 commented:
@ WoodsterDK : do you get any error when initializing the the component? It **should** load all entities automatically on startup: 

```
    async def async_start(self):
        """Start velux component."""
        await self.pyvlx.load_scenes()
        await self.pyvlx.load_nodes()
```

Julius2342 committed:
bumped version

Julius2342 commented:
@WoodsterDK : may you retry?

WoodsterDK commented:
> @WoodsterDK : may you retry?

Will test tomorrow...
Could it be caused by a combination with the switches I have? - the nodeIDs for my covers might not come in a numerical order due to the switches.
Do you lookup in the node array according to the nodeID or by an index found by searching for the nodeID?

Julius2342 commented:
@WoodsterDK : When retrieving the scenes, my old implementation was waiting for the last `FrameGetSceneListNotification` with `remaining_scenes==0`. If in your implementation no scenes are configured, this frame is never sent, the whole thing runs into a timeout and an exception is raised that the scene list may not be retrieved.

Julius2342 commented:
@WoodsterDK : And you are right, there was another bug with accessing the nodes via index-id instead of node_id (which makes more sense).

Julius2342 commented:
@WoodsterDK : And you are right, there was another bug with accessing the nodes via index-id instead of node_id (which makes more sense).

Julius2342 commented:
@WoodsterDK : And you are right, there was another bug with accessing the nodes via index-id instead of node_id (which makes more sense).

WoodsterDK commented:
> @WoodsterDK : And you are right, there was another bug with accessing the nodes via index-id instead of node_id (which makes more sense).

OK - just tried running with the updated library, and my covers seems to added as entities - GREAT.
I can control them, but it seems the position is not being correctly reflected.

Furthermore it could be nice to add the support of the feature cover.stop_cover in the component.

If I look in the log every 10 second there is this:
![image](https://user-images.githubusercontent.com/12181717/50273059-d71a9000-0439-11e9-9e9d-157386aa0741.png)
Even if there is no movement of the covers - do you poll for states from the products, or is it only the internal values that are being "broadcastet"?

Julius2342 commented:
what do you mean with `not being correctly reflected.` . I'dont know what the correct correlation is. 

WRT cover.stop, do you know which Frame to send?

No i do not poll. I activated the house-monitor. During movement they broadcast their position from time to time. Does this stop after the covers reached their final position?

WoodsterDK commented:
@Julius2342 Please skip the comment regarding the 10 sec. update
I forgot that I had a script running in the background - and your module with the Home Monitor reflected that update also..... DOH :)

Julius2342 commented:
regarding the slider: the internal position is only updated after the windows have reached their final position. Im thinking of some fancy intermediate prediction (bc the KLF 200 is responding with some time how long it will take to reach the target position, but i think we should do this with step 2. Same for stop covers.)

WoodsterDK commented:
I don't think the internal position is updated - my cover has reached the final position - current position = target position, but the slider hasn't changed

WoodsterDK commented:
Regarding the position - is register_device_updated_cb in node ever called?

Julius2342 committed:
added cover stop

Julius2342 committed:
bumped version of pyvlx

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into new-velux-api

Julius2342 commented:
@WoodsterDK: may you retry? I added `stop`. I also fixed the bug with the wrong node_id.

In the library itself i added also switches (not added to hass yet), may you give a try?

WoodsterDK commented:
@Julius2342 Great work - a couple of things:
- notification of position seems to be working, if the slider is being used.
- stop works
- position is being messed up when the movement is stopped by pressing "Stop"

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into new-velux-api

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into new-velux-api

jpearce73 commented:
Thanks for this work, I'd love to try it out.
I just upgraded homeassistant to 0.85.1. I found the components/velux.py and discovered its still the old version. Where do I learn how to pull pre-release change set like this?

WoodsterDK commented:
@Julius2342 have you had a chance to look at:

notification of position seems to be working, if the slider is being used.
stop works
position is being messed up when the movement is stopped by pressing "Stop"

Julius2342 commented:
@WoodsterDK : Yes, started a small refactoring but then I was interrupted. Will try to continue soon.

apeeters commented:
@WoodsterDK I did not experience that problem when pressing stop.

WoodsterDK commented:
@apeeters if you operate the cover, press stop close the ui card and open it again, the slider is positioned in a wrong position

apeeters commented:
@WoodsterDK @Julius2342 I was able to reproduce this. It does not always occur and is caused by negative position values. Initially all my covers (closed) have position -23, after moving and stopping I sometimes get -5.

l-mb commented:
Very interested in seeing this merged, let me know if I can test via a beta channel/repo for hass.io.

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into new-velux-api

Julius2342 committed:
Bumped version to 0.2.8

WoodsterDK commented:
@Julius2342 Sure thing - will see if I can fit in try today

marchingphoenix commented on homeassistant/components/cover/velux.py:
```diff
@@ -0,0 +1,96 @@
+    @callback
+    def async_register_callbacks(self):
+        """Register callbacks to update hass after device was changed."""
+        async def after_update_callback(device):
```
Question, should the `async def` be outside of the `def`? I'm not 100% familiar with this yet.
example: [cover.template.pyL201](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/cover/template.py#L201)

rytilahti replied:
I think that's okay, as it is just passed as a callback (which I assume gets awaited later on). The @callback marks this as something that does not do I/O (for optimization), see: https://developers.home-assistant.io/docs/en/asyncio_categorizing_functions.html#the-callback-function

WoodsterDK commented:
> @WoodsterDK may you retry?

@Julius2342 - just tested, with the reference to 'pyvlx==0.2.8' it seems to be working as expected... COOL :)

rytilahti commented:
Added the 'breaking change' label as according to the initial descriptions this will be such for existing users.

rytilahti commented on homeassistant/components/velux.py:
```diff
@@ -54,8 +54,10 @@ def __init__(self, hass, config):
             host=host,
-            password=password)
+            password=password,
+            log_frames=True)
```
What does this option do as it sounds like a debugging option, and why is it wanted here? Otherwise (considering the reports that this is working on real devices) this should be ready to be merged.

Julius2342 replied:
I removed this parameter!

Julius2342 committed:
removed log_frames parameter

rytilahti commented:
Thanks! :+1: I bolded the need to mention this in the changelog (inside description), as I don't think we have a label to highlight specific parts at the moment.

MartinHjelmare commented on homeassistant/components/cover/velux.py:
```diff
@@ -0,0 +1,96 @@
+
+    async def async_set_cover_position(self, **kwargs):
+        """Move the cover to a specific position."""
+        if ATTR_POSITION in kwargs:
```
This is already validated by the service schema.

MartinHjelmare reviewed (commented):
Good! Just a small comment.
